### ISSUE TITLE
Solution becomes flat while using a large timespan

### ISSUE THREAD

**Issue by AliCanter** [SciML/DiffEqJump.jl#177]
When solving a `JumpProblem` during large time spans, at one point the solution becomes just flat.
I tried with `SSAStepper()` and `FunctionMap()` and both solvers give similar results.
There are no warnings and the retval is `:Default` (`:Success` when using `FunctionMap()`).

A minimal example of this behavior is the following (adapted from [Catalyst.jl examples](https://catalyst.sciml.ai/dev/tutorials/basic_examples/)):
```julia
using Catalyst
using DiffEqJump
using Plots

rs = @reaction_network begin
  c1, X --> 2X
  c2, X --> 0
  c3, 0 --> X
end c1 c2 c3

p = (1., 2., 50.)
dt = 1.
tspan = (0., 2e6)
u0 = [5]

dprob = DiscreteProblem(u0, tspan, p)
jprob = JumpProblem(dprob, Direct(), rs, save_positions=(false, false))
sol   = solve(jprob, SSAStepper(), saveat=tspan[1]:dt:tspan[2])
plot(sol, lw=2, title="Birth-Death Process")
```

This happens for almost every system I tried, and happens earlier if there are a lot of jumps, so it may be a limit in the number of jumps.

**Comment by isaacsas** [comment]
Can reproduce with multiple SSA methods and when saving all time points.

**Comment by isaacsas** [comment]
Have you ever had this occur when not using `save_positions` or `saveat`? I can't check them for this example as there are too many jumps, but it would be helpful in debugging.

**Comment by AliCanter** [comment]
I don't think so, from what I tried the start of the plateau seems correlated with the amount of jumps that I am able save in RAM.

**Comment by isaacsas** [comment]
I think the issue is floating point error. It appears that right after this line:
https://github.com/SciML/DiffEqJump.jl/blob/b14226d7729fcf06cb1d5ecc4a05a755edfee5ea/src/aggregators/direct.jl#L57

We can have that 
```julia
p.next_jump_time == t
```
is true.

**Comment by isaacsas** [comment]
What seems to be happening is sometimes the time till the next jump is very small, on the order of 1e-13. It is just given by
```julia
ttnj = randexp() / sum_rate
```
In these cases the total propensity for the next reaction, `sum_rate`, is a couple hundred, so perfectly reasonably sized. So this is just coming from a very small value from `randexp` (relative to the current `t` value). Then 
```julia
p.next_jump_time = t + ttnj 
```
leaves `p_next_jump_time == t`. 

I'm not sure yet what the best approach to handle this is. We could put in a test that if this occurs we sample again, but that seems worrisome as a fix since we'd be skipping over legitimate jumps. I guess we could consider storing the next jump time in higher precision. I'm not sure if that would affect performance significantly.

"Efficient Formulations for Exact Stochastic Simulation of Chemical Systems" by Sean Mauch and Mark Stalzer suggests storing a time offset and epoch, using the offset internally, and once it becomes sufficiently large using it to increment the epoch while setting the offset back to zero. That's probably a better approach, but is going to require revamping all the SSAs...

@ChrisRackauckas any thoughts?

**Comment by isaacsas** [comment]
So I can confirm that just naively recalculating `ttnj` when it is too small gets rid of the issue, but I have to think a bit about what the right way to handle this is.

**Comment by ChrisRackauckas** [comment]
I think we need to just immediately do an event if the step is smaller than floating point.

**Comment by isaacsas** [comment]
We would probably need to then modify SSAStepper and FunctionMap to allow for multiple events at a given time.

**Comment by ChrisRackauckas** [comment]
I think it needs to be changed in the `affect!` that if time is small enough just immediately do a jump and recalculate.

**Comment by isaacsas** [comment]
Couldn’t that mess up discrete event handling by jumping past when the event should be applied?

**Comment by ChrisRackauckas** [comment]
No, I am saying that when you set the times for the new event then check if the time is < floating point eps, and if it is, just immediately do the next effect, take new random numbers, etc. Since it's all in the `affect!` function it doesn't matter.

**Comment by isaacsas** [comment]
Just to be clear, by event I meant discrete callback.

**Comment by isaacsas** [comment]
I agree with what you are saying in the absence of callbacks.

**Comment by ChrisRackauckas** [comment]
I don't see where the issue is. I'm just thinking that if the rate is high enough it should be simultaneous and you do more than one event at that time (and keep doing events until you get a small enough rate to resolve it in the near future).

**Comment by isaacsas** [comment]
What if a callback should go off when a species hits a given level? Wouldn’t you now potentially go past that and possibly miss executing the callback?

**Comment by isaacsas** [comment]
@AliCanter we'll have fix shortly, but another possible option is you can make `tspan` contain `Double64`s from DoubleFloat.jl. That seems to give enough precision to avoid the issue on this example. There will be a noticeable speed hit for the simulation though...
